File: argentum-api/src/main/java/net/devium/argentum/rest/model/response/ProductResponseMeta.java
package net.devium.argentum.rest.model.response;

import net.devium.argentum.jpa.ProductEntity;

import java.math.BigDecimal;

public class ProductResponseMeta {
    private final long id;
    private final String name;
    private final BigDecimal price;
    private final Long category;
    private final boolean legacy;

    private ProductResponseMeta(long id, String name, BigDecimal price, Long category, boolean legacy) {
        this.id = id;
        this.name = name;
        this.price = price;
        this.category = category;
        this.legacy = legacy;
    }

    public static ProductResponseMeta from(ProductEntity product) {
        return new ProductResponseMeta(
                product.getId(),
                product.getName(),
                product.getPrice(),
                product.getCategory() != null ? product.getCategory().getId() : null,
                product.isLegacy()
        );
    }

    public long getId() {
        return id;
    }

    public String getName() {
        return name;
    }

    public BigDecimal getPrice() {
        return price;
    }

    public Long getCategory() {
        return category;
    }

    public boolean isLegacy() {
        return legacy;
    }
}

File: argentum-ui/src/app/common/rest-service/response/product-response-meta.ts
import { Product } from '../../model/product';
import { Category } from '../../model/category';

export class ProductResponseMeta {
  id: number;
  name: string;
  price: number;
  category: number;
  legacy: boolean;
}

export function toProductMeta(response: ProductResponseMeta, categories: Category[]): Product {
  return {
    id: response.id,
    name: response.name,
    price: response.price,
    category: categories.find(category => category.id == response.category),
    legacy: response.legacy,
    ranges: null
  };
}
